server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java

```diff
@@ -84,10 +115,37 @@ MappedFieldType get(String field) {
         if (fieldType != null) {
             return fieldType;
         }
+        return getDynamicField(field);
+    }
+
+    // Check if the given field corresponds to a dynamic key mapper of the
+    // form 'path_to_field.path_to_key'. If so, returns a field type that
+    // can be used to perform searches on this field. Otherwise returns null.
+    private MappedFieldType getDynamicField(String field) {
+        if (dynamicFieldTypes.isEmpty()) {
+            // no parent fields defined
+            return null;
+        }
+        int dotIndex = -1;
+        int fieldDepth = -1;
+
+        while (true) {
```

Comment:
For some reason this loop makes me nervous that we may do more work than needed. Effectively we could stop once we encounter an object. But maybe that should not be a concern.

Reply:
> Effectively we could stop once we encounter an object

Not quite - you can have a dynamic field nested inside an object.  That's why we calculate the `maxParentPathDots` field, because once you've got past that you know that there are no dynamic roots that could match the path.

Reply:
but you go backwards analyzing the path, and you can't have a dynamic field pointing to an object, right? so once you find an object, you should be done and there is no need to look at its parent and so on? Am I missing something?

Reply:
My comment is off, because when you look up field types, you can not find an object :) so I think this is a non-issue, like I said I am nervous that we go ahead and analyze the path when it's not needed, but I am not sure that would be a problem and how to avoid it.